openssl/src/ssl/mod.rs

```diff
@@ -307,6 +307,39 @@ extern "C" fn raw_verify<F>(preverify_ok: c_int, x509_ctx: *mut ffi::X509_STORE_
     }
 }
 
+extern "C" fn raw_psk<F>(ssl: *mut ffi::SSL,
+                         hint: *const c_char,
+                         identity: *mut c_char,
+                         _max_identity_len: c_uint,
+                         psk: *mut c_uchar,
+                         _max_psk_len: c_uint) -> c_uint
+    where F: Fn(&mut SslRef, &str) -> Result<(String, Vec<u8>), ErrorStack> + Any + 'static + Sync + Send
```

Comment:
I wrapped it in a more Rust-friendly API because it seems like it would be error-prone to expose it at that level. Specifically, `identity` needs to be null-terminated. (Passing `hint` as an `Option` is much better though, I don't know why I didn't think of that!)

I pushed my code that's using this to GitHub so you can see what it looks like in practice with the current implementation:
https://github.com/luser/steamstreaming/blob/004611fdc8db949020e4e25011df361f01e1e39a/src/lib.rs#L86

I'm happy to do whatever you think is best here, though.